src/passport-saml/saml.ts

```diff
@@ -108,99 +108,80 @@ async function promiseWithNameID(nameid: Node): Promise<NameID> {
+  constructor(ctorOptions: SamlConfig) {
+    this.options = this.initialize(ctorOptions);
     this.cacheProvider = this.options.cacheProvider;
   }
-  initialize(options: Partial<SamlOptions>): SamlOptions {
-    if (!options) {
-      options = {};
+
+  initialize(ctorOptions: SamlConfig): SamlOptions {
+    if (!ctorOptions) {
+      throw new TypeError("SamlOptions required on construction");
     }
 
-    if (options.privateCert) {
+    if (ctorOptions.privateCert) {
       console.warn("options.privateCert has been deprecated; use options.privateKey instead.");
 
-      if (!options.privateKey) {
-        options.privateKey = options.privateCert;
+      if (!ctorOptions.privateKey) {
+        ctorOptions.privateKey = ctorOptions.privateCert;
       }
     }
 
-    if (Object.prototype.hasOwnProperty.call(options, "cert") && !options.cert) {
-      throw new Error("Invalid property: cert must not be empty");
-    }
-
-    if (!options.path) {
-      options.path = "/saml/consume";
-    }
-
-    if (!options.host) {
-      options.host = "localhost";
-    }
-
-    if (!options.issuer) {
-      options.issuer = "onelogin_saml";
-    }
-
-    if (options.identifierFormat === undefined) {
-      options.identifierFormat = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress";
-    }
-
-    if (options.authnContext === undefined) {
-      options.authnContext = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport";
-    }
-
-    if (!Array.isArray(options.authnContext)) {
-      options.authnContext = [options.authnContext];
-    }
-
-    if (!options.acceptedClockSkewMs) {
-      // default to no skew
-      options.acceptedClockSkewMs = 0;
-    }
-
-    if (!options.validateInResponseTo) {
-      options.validateInResponseTo = false;
-    }
-
-    if (!options.requestIdExpirationPeriodMs) {
-      options.requestIdExpirationPeriodMs = 28800000; // 8 hours
-    }
-
-    if (!options.cacheProvider) {
-      options.cacheProvider = new InMemoryCacheProvider({
-        keyExpirationPeriodMs: options.requestIdExpirationPeriodMs,
-      });
-    }
-
-    if (!options.logoutUrl) {
-      // Default to Entry Point
-      options.logoutUrl = options.entryPoint || "";
-    }
-
-    // sha1, sha256, or sha512
-    if (!options.signatureAlgorithm) {
-      options.signatureAlgorithm = "sha1";
-    }
-
-    /**
-     * List of possible values:
-     * - exact : Assertion context must exactly match a context in the list
-     * - minimum:  Assertion context must be at least as strong as a context in the list
-     * - maximum:  Assertion context must be no stronger than a context in the list
-     * - better:  Assertion context must be stronger than all contexts in the list
-     */
-    if (
-      !options.RACComparison ||
-      ["exact", "minimum", "maximum", "better"].indexOf(options.RACComparison) === -1
-    ) {
-      options.RACComparison = "exact";
-    }
+    const options = {
+      ...ctorOptions,
```

Comment:
I'm not sure I understand everything that's going on.
Why not simply have something like:
```
const DEFAULTS = {
   passive: false,
   disableRequestedAuthnContext: false,
   // all the default values
}

const options = {...DEFAULTS, ...ctorOptions};
```

Note that this can still cause problems with the user explicitly passes an option to null or undefined (like `new SAML({passive: undefined})` ) so some kind of cleanup might be needed beforehand (for JS users - TS users should not have this issue thanks to typing)

Something like
```
// remove all the undefined, and the null for non nullable options.
Object.keys(ctorOptions).forEach((key) => {
   if(ctorOptions[key] === undefined) delete ctorOptions[key];
   if(ctorOptions[key] === null && key !== 'identifierFormat' && key !== 'authnContext') delete ctorOptions[key];
})
```

Enumerating all the properties and their default with `??` as you did works too, but in this case I would not add the `...ctorOptions` at the beginning, which IMO is confusing (and maybe just explicit all properties that don't have any default, if any)

Reply:
Putting in `...ctorOptions` like is done is so that all properties don't have to be explicitly specified, only ones where there are some defaults or special handling. 

Generally, I see what you are after here, but I feel like this logic might be easier to follow than the hoops you describe. It also makes it easier to extend in one place instead of having to go to multiple places. ("I have to remember that if `null` is a valid value, to go to this other block to explicitly allow that.")